display_list/display_list_canvas_unittests.cc

```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
+
+  void render(SkCanvas* canvas) {
+    get_display_list()->RenderTo(canvas, opacity);
+  }
+
+  sk_sp<DisplayList> get_display_list() {
+    if (!display_list) {
+      DisplayListBuilder builder(SkRect::MakeWH(width, height));
+      dl_setup(builder);
+      setup_matrix = builder.getTransform();
+      setup_clip_bounds = builder.getDestinationClipBounds().roundOut();
+      dl_render(builder);
+      dl_restore(builder);
+      display_list = builder.Build();
+    }
+    return display_list;
+  }
 };
 
 class RenderEnvironment {
  public:
-  static RenderEnvironment Make565() {
-    return RenderEnvironment(
-        SkImageInfo::Make({1, 1}, kRGB_565_SkColorType, kOpaque_SkAlphaType));
+  RenderEnvironment(const DlSurfaceProvider* provider, PixelFormat format)
+      : provider_(provider), format_(format) {
+    if (provider->supports(format)) {
+      main_surface_ =
+          provider->MakeOffscreenSurface(kTestWidth, kTestHeight, format);
+      alt_surface_ = provider->MakeOffscreenSurface(kTestWidth * 2,
+                                                    kTestHeight * 2, format);
+    }
   }
 
-  static RenderEnvironment MakeN32() {
-    return RenderEnvironment(SkImageInfo::MakeN32Premul(1, 1));
+  static RenderEnvironment Make565(const DlSurfaceProvider* provider) {
+    return RenderEnvironment(provider, PixelFormat::k565_PixelFormat);
   }
 
-  std::unique_ptr<RenderSurface> MakeSurface(
-      const DlColor bg = DlColor::kTransparent(),
-      int width = kTestWidth,
-      int height = kTestHeight) const {
-    sk_sp<SkSurface> surface =
-        SkSurface::MakeRaster(info_.makeWH(width, height));
-    surface->getCanvas()->clear(bg);
-    return std::make_unique<RenderSurface>(surface);
+  static RenderEnvironment MakeN32(const DlSurfaceProvider* provider) {
+    return RenderEnvironment(provider, PixelFormat::kN32Premul_PixelFormat);
   }
 
   void init_ref(CvRenderer& cv_renderer, DlColor bg = DlColor::kTransparent()) {
-    init_ref([=](SkCanvas*, SkPaint&) {}, cv_renderer,
-             [=](DisplayListBuilder&) {}, bg);
+    init_ref(kEmptyCvSetup, cv_renderer, kEmptyDlRenderer, bg);
   }
 
   void init_ref(CvSetup& cv_setup,
                 CvRenderer& cv_renderer,
                 DlRenderer& dl_setup,
                 DlColor bg = DlColor::kTransparent()) {
-    ref_canvas()->clear(bg);
+    CvRenderJob job = {
+        .bg = bg,
+        .cv_setup = cv_setup,
+        .cv_render = cv_renderer,
+    };
+    ref_result_ = getResult(job);
     dl_setup(ref_attr_);
-    SkPaint paint;
-    cv_setup(ref_canvas(), paint);
-    ref_matrix_ = ref_canvas()->getTotalMatrix();
-    ref_clip_ = ref_canvas()->getDeviceClipBounds();
-    cv_renderer(ref_canvas(), paint);
-    ref_pixmap_ = ref_surface_->pixmap();
   }
 
-  const SkImageInfo& info() const { return info_; }
-  SkCanvas* ref_canvas() { return ref_surface_->canvas(); }
+  template <typename J>
+  std::unique_ptr<RenderResult> getResult(J& job) const {
```

Comment:
I may be missing something but I am not sure why this needs to be templated if the jobs can subclass a common struct.

Reply:
Struct initializers prevent virtual methods.

Reply:
Fixed as mentioned above.